// reshaped/components/Overlay/tests/Overlay.stories.js
import { Example } from "../../../utilities/storybook/index.js";
import Overlay from "../index.js";
import Button from "../../Button/index.js";
import useToggle from "../../../hooks/useToggle.js";
export default {
    title: "Utilities/Overlay",
    component: Overlay,
    parameters: {
        iframe: {
            url: "https://reshaped.so/docs/components/overlay",
        },
    },
};
export const base = () => {
    const baseToggle = useToggle(true);
    const transparentToggle = useToggle(false);
    return (<Example>
			<Example.Item title="locks scroll">
				<Button onClick={() => baseToggle.activate()}>Open overlay</Button>
				<Overlay active={baseToggle.active} onClose={() => baseToggle.deactivate()}>
					Overlay content
				</Overlay>
				<div style={{ height: 1000 }}/>
			</Example.Item>

			<Example.Item title="transparent, doesn't lock scroll">
				<Button onClick={() => transparentToggle.activate()}>Open overlay</Button>
				<Overlay active={transparentToggle.active} onClose={() => transparentToggle.deactivate()} transparent>
					Overlay content
				</Overlay>
				<div style={{ height: 1000 }}/>
			</Example.Item>
		</Example>);
};
